### PR TITLE
Parse "Group Id" from encoded Vulkan debug labels.

### PR DESCRIPTION
**Pull request by florian-kuebler**
In DXVK, we insert Vulkan debug labels with the following encoding:
'DXVK__vkFunctionName#GROUP_ID', where 'GROUP_ID' is the group
id to be used.

This parses the labels and sets the group id field if possible.

There are already tests for GpuQueueSubmissionProcessor in
the CaptureEventProcessorTest, but I believe it is better
to have those special tests closer to the actual class.

We should in the future, add some more tests for the corner
cases in the new test file.

Test: Take a capture on a DXVK sample, run tests.

### CODE REVIEW

File: src/CaptureClient/GpuQueueSubmissionProcessor.cpp [hunk 1]
```diff
@@ -412,7 +415,19 @@ std::vector<TimerInfo> GpuQueueSubmissionProcessor::ProcessGpuDebugMarkers(
...
-    marker_timer.set_user_data_key(completed_marker.text_key());
+
+    // We have special handling for DXVK instrumentation that extracts the encoded group_id from
+    // the label's text. The encoding is:
+    // 'DXVK__vkFunctionName#GROUP_ID', where 'GROUP_ID' is the group id.
+    uint64_t text_key = completed_marker.text_key();
+    marker_timer.set_user_data_key(text_key);
+    std::smatch dxvk_group_id_match;
+    CHECK(string_intern_pool.contains(text_key));
+    if (std::regex_match(string_intern_pool.at(text_key), dxvk_group_id_match,
```
Comment (ronaldfw) [inline src/CaptureClient/GpuQueueSubmissionProcessor.cpp:426]: On first glance, using regular expressions here seems to be overkill. You need to make sure the start of the string is "DXVK" and then find the last occurrence of "#".

Comment (florian-kuebler) [reply src/CaptureClient/GpuQueueSubmissionProcessor.cpp:426]: But isn't that safer and more readable?

Comment (ronaldfw) [reply src/CaptureClient/GpuQueueSubmissionProcessor.cpp:426]: I don't think using regular expressions is more readable, but that's more a personal preference and others may disagree. 

The one part where it would currently be safer is making sure that the part after "#" is actually just digits. But you could use absl::SimpleAtoi and check the return value. Not sure where we stand regarding use of stoull and the exceptions it can throw. 

The question is more: Why use something that is much more powerful than necessary? IMO, that's unnecessary complexity. I suppose <regex> probably doesn't have that many issues, but bugs in <regex> are still much more likely than in a simple string search. 

Another thing worth looking at: Is this performance critical code?

File: src/CaptureClient/GpuQueueSubmissionProcessor.cpp [hunk 2]
```diff
@@ -412,7 +415,19 @@ std::vector<TimerInfo> GpuQueueSubmissionProcessor::ProcessGpuDebugMarkers(
       color->set_blue(static_cast<uint32_t>(completed_marker.color().blue() * 255.f));
       color->set_alpha(static_cast<uint32_t>(completed_marker.color().alpha() * 255.f));
     }
-    marker_timer.set_user_data_key(completed_marker.text_key());
+
+    // We have special handling for DXVK instrumentation that extracts the encoded group_id from
+    // the label's text. The encoding is:
+    // 'DXVK__vkFunctionName#GROUP_ID', where 'GROUP_ID' is the group id.
+    uint64_t text_key = completed_marker.text_key();
```
Comment (danielfenner) [inline src/CaptureClient/GpuQueueSubmissionProcessor.cpp:422]: Nit: Merely for readability

`const uint64_t text_key = completed_marker.text_key();`

File: src/CaptureClient/GpuQueueSubmissionProcessor.cpp (hunk 1 above)
Comment (danielfenner) [reply src/CaptureClient/GpuQueueSubmissionProcessor.cpp:426]: If this is an performance issue I'd agree and try something like
```
size_t found = string_intern_pool.at(text_key).find_last_of("#");
... absl::SimpleAtoi (string_intern_pool.at(text_key).substr(found+1));
```
Otherwise the regex does not bother me. I'm not sure how the regex would react to a string with two "#" in it though. Does it bind to the first or the second # in "asd#111#222". May be worth testing.

Comment (ronaldfw) [reply src/CaptureClient/GpuQueueSubmissionProcessor.cpp:426]: > I'm not sure how the regex would react to a string with two "#" in it though.

I would argue that this question alone is an argument that using "find_last_of" is better and more readable.

Comment (florian-kuebler) [reply src/CaptureClient/GpuQueueSubmissionProcessor.cpp:426]: So a version without the regex would look like this:

```
    if (text.find_first_of("DXVK__") != std::string::npos) {
      const size_t group_id_index = text.find_last_of('#') + 1;
      uint64_t group_id = 0;
      if (group_id_index != std::string::npos &&
          absl::SimpleAtoi(text.substr(group_id_index), &group_id)) {
        marker_timer.set_group_id(group_id);
      }
    }
```

Vs.:
```
    std::smatch dxvk_group_id_match;
    if (std::regex_match(text, dxvk_group_id_match, std::regex("DXVK__vk.*#([0-9]+)")) &&
        dxvk_group_id_match.size() == 2) {
      uint64_t group_id = 0;
      if (absl::SimpleAtoi(dxvk_group_id_match[1].str(), &group_id)) {
        marker_timer.set_group_id(group_id);
      }
    }
```

So yeah, I don't know, I prefer the regex, as it gives me basically directly both things to look at -- the start and the encoded end. But I agree this is personal preference, I am just very used to working with regexes, so I am biased here.

Regarding performance: 1. This is on the client, so still a concern, but less as if it would be in the service.
And yes, if there are a lot of debug marker, then this is on a hot path. Every debug marker needs to get processed here.

Performance of the two implementations would for sure required to be measured if it would really make a difference.

From inspecting the code, the case were we have few debug labels that contains the "DXVK__" prefix, the regex implementation would be probably faster, as it can early exit and there is no proper "starts_with" in cpp for the other implementation (sure we can do this with substring, but then we also need to add the bounds checks and the code becomes less readable).

For the case that the label actually is a DXVK__ label, the other implementation is *probably* faster. Though, complexity-wise they are the same.

Comment (florian-kuebler) [reply src/CaptureClient/GpuQueueSubmissionProcessor.cpp:426]: >>Otherwise the regex does not bother me. I'm not sure how the regex would react to a string with two "#" in it though. Does it bind to the first or the second # in "asd#111#222". May be worth testing.

Yes, regex's are greedy (and thus proceed in the string as long as possible), `.*?` would be the non-greedy version.

>I would argue that this question alone is an argument that using "find_last_of" is better and more readable.

Sure, I am fine with changing this.

File: src/CaptureClient/GpuQueueSubmissionProcessor.cpp (hunk 2 above)
Comment (florian-kuebler) [reply src/CaptureClient/GpuQueueSubmissionProcessor.cpp:422]: Done

File: src/CaptureClient/GpuQueueSubmissionProcessor.cpp (hunk 1 above)
Comment (florian-kuebler) [reply src/CaptureClient/GpuQueueSubmissionProcessor.cpp]: Changed it.

Comment (ronaldfw) [reply src/CaptureClient/GpuQueueSubmissionProcessor.cpp]: Once we use C++20, we can switch to using `starts_with`.

File: src/CaptureClient/GpuQueueSubmissionProcessor.cpp [hunk 3]
```diff
@@ -412,7 +413,23 @@ std::vector<TimerInfo> GpuQueueSubmissionProcessor::ProcessGpuDebugMarkers(
...
-    marker_timer.set_user_data_key(completed_marker.text_key());
+
+    const uint64_t text_key = completed_marker.text_key();
+    marker_timer.set_user_data_key(text_key);
+
+    // We have special handling for DXVK instrumentation that extracts the encoded group_id from
+    // the label's text. The encoding is:
+    // 'DXVK__vkFunctionName#GROUP_ID', where 'GROUP_ID' is the group id.
+    CHECK(string_intern_pool.contains(text_key));
+    const std::string& text = string_intern_pool.at(text_key);
```
Comment (ronaldfw) [inline src/CaptureClient/GpuQueueSubmissionProcessor.cpp:424]: Would it make sense to pull this out as a testable function and fully unit test this piece of code?

Comment (ronaldfw) [reply src/CaptureClient/GpuQueueSubmissionProcessor.cpp:424]: It seems right now we are only testing a single well-formed label.

**Review by ronaldfw — APPROVED** [review]
LGTM, one small thing to consider, but up to you

File: src/CaptureClient/GpuQueueSubmissionProcessor.cpp (hunk 3 above)
Comment (florian-kuebler) [reply src/CaptureClient/GpuQueueSubmissionProcessor.cpp:424]: Makes sense. Will do.

Comment (florian-kuebler) [reply src/CaptureClient/GpuQueueSubmissionProcessor.cpp:424]: Done!

**Review by danielfenner — APPROVED** [review]
lgtm!

**Review by ronaldfw — APPROVED** [review]
LGTM
